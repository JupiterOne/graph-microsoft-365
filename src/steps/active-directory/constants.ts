// Step IDs
export const STEP_ACCOUNT = 'account';
export const STEP_GROUPS = 'groups';
export const STEP_GROUP_MEMBERS = 'group-members';
export const STEP_USERS = 'users';

// Graph objects
export const ACCOUNT_ENTITY_TYPE = 'microsoft_365_account';
export const ACCOUNT_ENTITY_CLASS = 'Account';

export const GROUP_ENTITY_TYPE = 'microsoft_365_user_group';
export const GROUP_ENTITY_CLASS = 'UserGroup';

export const USER_ENTITY_TYPE = 'microsoft_365_user';
export const USER_ENTITY_CLASS = 'User';

/**
 * The entity type used for members of groups which are not one of the ingested
 * directory objects.
 */
export const GROUP_MEMBER_ENTITY_TYPE = 'microsoft_365_group_member';

/**
 * The entity class used for members of groups which are not one of the ingested
 * directory objects.
 */
export const GROUP_MEMBER_ENTITY_CLASS = 'User';

export const ACCOUNT_GROUP_RELATIONSHIP_TYPE =
  'microsoft_365_account_has_group';
export const ACCOUNT_USER_RELATIONSHIP_TYPE = 'microsoft_365_account_has_user';
export const GROUP_MEMBER_RELATIONSHIP_TYPE = 'microsoft_365_group_has_member';
