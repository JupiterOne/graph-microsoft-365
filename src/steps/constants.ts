export const IngestionSources = {
  GROUPS: 'groups',
  USERS: 'users',
  APPLICATIONS: 'applications',
  DEVICES: 'devices',
  COMPLIANCE_POLICIES: 'compliance_policies',
  DEVICE_CONFIGURATIONS: 'device_configurations',
};
