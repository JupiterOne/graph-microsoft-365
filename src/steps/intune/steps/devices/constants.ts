import {
  RelationshipClass,
  StepEntityMetadata,
  StepRelationshipMetadata,
} from '@jupiterone/integration-sdk-core';
import { entities as activeDirectoryEntities } from '../../../active-directory';

export const steps: { [k: string]: string } = {
  FETCH_DEVICES: 'managed-devices',
};

export const entities: { [k: string]: StepEntityMetadata } = {
  DEVICE: {
    resourceName: 'Managed Device',
    _type: 'intune_managed_device',
    _class: ['Host', 'Device'],
  },
};

export const relationships: { [k: string]: StepRelationshipMetadata } = {
  USER_HAS_DEVICE: {
    _type: 'microsoft_365_user_has_intune_managed_device',
    sourceType: activeDirectoryEntities.USER._type,
    _class: RelationshipClass.HAS,
    targetType: entities.DEVICE._type,
  },
};
