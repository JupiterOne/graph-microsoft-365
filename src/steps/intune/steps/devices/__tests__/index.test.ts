import {
  createMockStepExecutionContext,
  Recording,
} from '@jupiterone/integration-sdk-testing';
import { setupAzureRecording } from '../../../../../../test/recording';
import { config } from '../../../../../../test/config';
import { fetchDevices } from '..';
import { entities } from '../../../constants';
import { fetchAccount, fetchUsers } from '../../../../active-directory';
import { ensureArray } from '../../../../../../test/ensureArray';
import { isEqual, omit } from 'lodash';

let recording: Recording;

afterEach(async () => {
  if (recording) {
    await recording.stop();
  }
});

describe('fetchDevices', () => {
  test('Without Active Directory', async () => {
    recording = setupAzureRecording({
      directory: __dirname,
      name: 'iterateManagedDevicesNoAD',
    });
    const context = createMockStepExecutionContext({ instanceConfig: config });

    await fetchDevices(context);

    const managedDevices = context.jobState.collectedEntities.filter((e) =>
      isEqual(e._class, ensureArray(entities.DEVICE._class)),
    );
    const managedDeviceRelationships = context.jobState.collectedRelationships.filter(
      (r) => {
        return r._type.includes(entities.DEVICE._type);
      },
    );

    // Check that we have devices
    expect(managedDevices.length).toBeGreaterThan(0);
    expect(
      managedDevices.map((d) => omit(d, 'userEmails')), // REDACTING the emails messes with the schema validation. Removing for this step.
    ).toMatchGraphObjectSchema({
      _class: ['Device'],
    });
    expect(managedDevices).toMatchSnapshot('managedDevices'); // intentionally the same snapshot as in the 'With Active Directory' test below

    // Check that relationship is mapped
    expect(managedDeviceRelationships.length).toBeGreaterThan(0);
    managedDeviceRelationships.forEach((r) =>
      expect(r).toMatchObject({
        _mapping: expect.any(Object),
      }),
    );
    expect(managedDeviceRelationships).toMatchSnapshot(
      'managedDeviceRelationshipsNoAD',
    );
  });

  test('With Active Directory', async () => {
    recording = setupAzureRecording({
      directory: __dirname,
      name: 'iterateManagedDevicesWithAD',
    });
    const context = createMockStepExecutionContext({ instanceConfig: config });

    await fetchAccount(context);
    await fetchUsers(context);
    await fetchDevices(context);

    const managedDevices = context.jobState.collectedEntities.filter((e) =>
      isEqual(e._class, ensureArray(entities.DEVICE._class)),
    );
    const managedDeviceRelationships = context.jobState.collectedRelationships.filter(
      (r) => {
        return r._type.includes(entities.DEVICE._type);
      },
    );

    // Check that we have devices
    expect(managedDevices.length).toBeGreaterThan(0);
    expect(
      managedDevices.map((d) => omit(d, 'userEmails')), // REDACTING the emails messes with the schema validation. Removing for this step.
    ).toMatchGraphObjectSchema({
      _class: ['Device'],
    });
    expect(managedDevices).toMatchSnapshot('managedDevices'); // intentionally the same snapshot as in the 'Without Active Directory' test above

    // Check that relationship is direct
    expect(managedDeviceRelationships.length).toBeGreaterThan(0);
    expect(managedDeviceRelationships).toMatchDirectRelationshipSchema({});
    expect(managedDeviceRelationships).toMatchSnapshot(
      'managedDeviceRelationshipsWithAD',
    );
  });
});
