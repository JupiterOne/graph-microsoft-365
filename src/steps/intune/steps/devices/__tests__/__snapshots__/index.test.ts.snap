// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchDevices With Active Directory: managedDeviceRelationshipsWithAD 1`] = `
Array [
  Object {
    "_class": "HAS",
    "_fromEntityKey": "001d35fc-dd2b-4d17-9ffc-3bafa265e7d9",
    "_key": "001d35fc-dd2b-4d17-9ffc-3bafa265e7d9|has|683dbff1-c8ff-4996-91ae-85484de46cfc",
    "_toEntityKey": "683dbff1-c8ff-4996-91ae-85484de46cfc",
    "_type": "azure_user_has_intune_managed_device",
    "displayName": "HAS",
  },
]
`;

exports[`fetchDevices With Active Directory: managedDevices 1`] = `
Array [
  Object {
    "BYOD": true,
    "_class": Array [
      "Device",
    ],
    "_key": "683dbff1-c8ff-4996-91ae-85484de46cfc",
    "_rawData": Array [
      Object {
        "name": "default",
        "rawData": Object {
          "aadRegistered": null,
          "activationLockBypassCode": null,
          "androidSecurityPatchLevel": "",
          "autopilotEnrolled": false,
          "azureADDeviceId": "683dbff1-c8ff-4996-91ae-85484de46cfc",
          "azureADRegistered": null,
          "azureActiveDirectoryDeviceId": "683dbff1-c8ff-4996-91ae-85484de46cfc",
          "chassisType": "unknown",
          "complianceGracePeriodExpirationDateTime": "2021-02-17T19:13:41Z",
          "complianceState": "noncompliant",
          "configurationManagerClientEnabledFeatures": null,
          "configurationManagerClientHealthState": null,
          "configurationManagerClientInformation": null,
          "deviceActionResults": Array [],
          "deviceCategoryDisplayName": "Unknown",
          "deviceEnrollmentType": "deviceEnrollmentManager",
          "deviceHealthAttestationState": null,
          "deviceName": "[REDACTED]",
          "deviceRegistrationState": "registered",
          "deviceType": "iPhone",
          "easActivated": false,
          "easActivationDateTime": "0001-01-01T00:00:00Z",
          "easDeviceId": "7CDT0MKDOH5FB0CKI5IVON0GJ8",
          "emailAddress": "[REDACTED]",
          "enrolledDateTime": "2021-02-17T16:24:45Z",
          "ethernetMacAddress": null,
          "exchangeAccessState": "none",
          "exchangeAccessStateReason": "none",
          "exchangeLastSuccessfulSyncDateTime": "0001-01-01T00:00:00Z",
          "freeStorageSpaceInBytes": 14481883136,
          "hardwareInformation": Object {
            "batteryChargeCycles": 0,
            "batteryHealthPercentage": 0,
            "batterySerialNumber": null,
            "cellularTechnology": null,
            "deviceFullQualifiedDomainName": null,
            "deviceGuardLocalSystemAuthorityCredentialGuardState": "running",
            "deviceGuardVirtualizationBasedSecurityHardwareRequirementState": "meetHardwareRequirements",
            "deviceGuardVirtualizationBasedSecurityState": "running",
            "freeStorageSpace": 0,
            "imei": "353249104017369",
            "ipAddressV4": null,
            "isEncrypted": false,
            "isSharedDevice": false,
            "isSupervised": false,
            "manufacturer": null,
            "meid": null,
            "model": null,
            "operatingSystemEdition": null,
            "operatingSystemLanguage": null,
            "operatingSystemProductType": 0,
            "osBuildNumber": null,
            "phoneNumber": null,
            "serialNumber": "C39ZQB9SN6XQ",
            "sharedDeviceCachedUsers": Array [],
            "subnetAddress": null,
            "subscriberCarrier": null,
            "totalStorageSpace": 0,
            "tpmSpecificationVersion": null,
            "wifiMac": null,
          },
          "iccid": null,
          "id": "683dbff1-c8ff-4996-91ae-85484de46cfc",
          "imei": "[REDACTED]",
          "isEncrypted": true,
          "isSupervised": false,
          "jailBroken": "False",
          "joinType": "unknown",
          "lastSyncDateTime": "2021-03-01T16:23:39Z",
          "lostModeState": "disabled",
          "managedDeviceName": "[REDACTED]",
          "managedDeviceOwnerType": "personal",
          "managementAgent": "mdm",
          "managementCertificateExpirationDate": "2022-02-15T19:56:28Z",
          "managementFeatures": "none",
          "managementState": "managed",
          "manufacturer": "Apple",
          "meid": "[REDACTED]",
          "model": "iPhone 11 Pro",
          "notes": null,
          "operatingSystem": "iOS",
          "osVersion": "14.4",
          "ownerType": "personal",
          "partnerReportedThreatState": "unknown",
          "phoneNumber": "[REDACTED]",
          "physicalMemoryInBytes": 0,
          "preferMdmOverGroupPolicyAppliedDateTime": "0001-01-01T00:00:00Z",
          "processorArchitecture": "unknown",
          "remoteAssistanceSessionErrorDetails": null,
          "remoteAssistanceSessionUrl": null,
          "requireUserEnrollmentApproval": null,
          "retireAfterDateTime": "0001-01-01T00:00:00Z",
          "roleScopeTagIds": Array [],
          "serialNumber": "[REDACTED]",
          "skuFamily": "",
          "skuNumber": 0,
          "specificationVersion": null,
          "subscriberCarrier": "Sprint",
          "totalStorageSpaceInBytes": 68719476736,
          "udid": null,
          "userDisplayName": "Michael Knoedel",
          "userId": "001d35fc-dd2b-4d17-9ffc-3bafa265e7d9",
          "userPrincipalName": "[REDACTED]",
          "usersLoggedOn": Array [
            Object {
              "lastLogOnDateTime": "2021-03-01T16:23:39.182Z",
              "userId": "001d35fc-dd2b-4d17-9ffc-3bafa265e7d9",
            },
          ],
          "wiFiMacAddress": "[REDACTED]",
          "windowsActiveMalwareCount": 0,
          "windowsRemediatedMalwareCount": 0,
        },
      },
    ],
    "_type": "intune_managed_device",
    "assetTag": "683dbff1-c8ff-4996-91ae-85484de46cfc",
    "category": "endpoint",
    "complianceState": "noncompliant",
    "createdOn": undefined,
    "deviceName": "[REDACTED]",
    "displayName": "[REDACTED]",
    "easDeviceId": false,
    "encrypted": true,
    "hardwareManufacturer": "Apple",
    "hardwareModel": "iPhone 11 Pro",
    "hardwareSerial": "C39ZQB9SN6XQ",
    "hardwareVendor": "Apple",
    "hardwareVersion": "iPhone 11 Pro",
    "iccid": null,
    "id": "683dbff1-c8ff-4996-91ae-85484de46cfc",
    "imei": "[REDACTED]",
    "jailBroken": false,
    "make": "Apple",
    "managed": true,
    "managementAgent": "mdm",
    "meid": "[REDACTED]",
    "model": "iPhone 11 Pro",
    "name": "[REDACTED]",
    "osDetails": "iOS version 14.4",
    "osName": "iOS",
    "osVersion": "14.4",
    "ownerType": "personal",
    "phoneNumber": "[REDACTED]",
    "physical": true,
    "registrationState": "registered",
    "serial": "[REDACTED]",
    "serialNumber": "[REDACTED]",
    "state": "managed",
    "supervised": false,
    "udid": null,
    "userDisplayName": "Michael Knoedel",
    "userEmails": Array [
      "[REDACTED]",
    ],
    "userId": "001d35fc-dd2b-4d17-9ffc-3bafa265e7d9",
    "username": "[REDACTED]",
    "version": "iPhone 11 Pro",
    "wifiMacAddress": "[REDACTED]",
  },
  Object {
    "BYOD": false,
    "_class": Array [
      "Device",
    ],
    "_key": "1f75329e-51fc-451c-bd40-c457337641a4",
    "_rawData": Array [
      Object {
        "name": "default",
        "rawData": Object {
          "aadRegistered": false,
          "activationLockBypassCode": null,
          "androidSecurityPatchLevel": "",
          "autopilotEnrolled": false,
          "azureADDeviceId": "00000000-0000-0000-0000-000000000000",
          "azureADRegistered": false,
          "azureActiveDirectoryDeviceId": "00000000-0000-0000-0000-000000000000",
          "chassisType": "unknown",
          "complianceGracePeriodExpirationDateTime": "0001-01-01T00:00:00Z",
          "complianceState": "unknown",
          "configurationManagerClientEnabledFeatures": null,
          "configurationManagerClientHealthState": null,
          "configurationManagerClientInformation": null,
          "deviceActionResults": Array [],
          "deviceCategoryDisplayName": "Unknown",
          "deviceEnrollmentType": "appleBulkWithoutUser",
          "deviceHealthAttestationState": null,
          "deviceName": "[REDACTED]",
          "deviceRegistrationState": "registered",
          "deviceType": "macMDM",
          "easActivated": false,
          "easActivationDateTime": "0001-01-01T00:00:00Z",
          "easDeviceId": "ApplC02DG307MD6Q",
          "emailAddress": "",
          "enrolledDateTime": "2021-02-17T16:38:03Z",
          "ethernetMacAddress": null,
          "exchangeAccessState": "none",
          "exchangeAccessStateReason": "none",
          "exchangeLastSuccessfulSyncDateTime": "0001-01-01T00:00:00Z",
          "freeStorageSpaceInBytes": 818191269888,
          "hardwareInformation": Object {
            "batteryChargeCycles": 0,
            "batteryHealthPercentage": 0,
            "batterySerialNumber": null,
            "cellularTechnology": null,
            "deviceFullQualifiedDomainName": null,
            "deviceGuardLocalSystemAuthorityCredentialGuardState": "running",
            "deviceGuardVirtualizationBasedSecurityHardwareRequirementState": "meetHardwareRequirements",
            "deviceGuardVirtualizationBasedSecurityState": "running",
            "freeStorageSpace": 0,
            "imei": "",
            "ipAddressV4": null,
            "isEncrypted": false,
            "isSharedDevice": false,
            "isSupervised": false,
            "manufacturer": null,
            "meid": null,
            "model": null,
            "operatingSystemEdition": null,
            "operatingSystemLanguage": null,
            "operatingSystemProductType": 0,
            "osBuildNumber": null,
            "phoneNumber": null,
            "serialNumber": "C02DG307MD6Q",
            "sharedDeviceCachedUsers": Array [],
            "subnetAddress": null,
            "subscriberCarrier": null,
            "totalStorageSpace": 0,
            "tpmSpecificationVersion": null,
            "wifiMac": null,
          },
          "iccid": null,
          "id": "1f75329e-51fc-451c-bd40-c457337641a4",
          "imei": "",
          "isEncrypted": true,
          "isSupervised": true,
          "jailBroken": "Unknown",
          "joinType": "unknown",
          "lastSyncDateTime": "2021-02-18T16:26:58Z",
          "lostModeState": "disabled",
          "managedDeviceName": "[REDACTED]",
          "managedDeviceOwnerType": "company",
          "managementAgent": "mdm",
          "managementCertificateExpirationDate": "2022-02-16T22:35:43Z",
          "managementFeatures": "none",
          "managementState": "retireIssued",
          "manufacturer": "Apple",
          "meid": "",
          "model": "MacBook Pro",
          "notes": null,
          "operatingSystem": "macOS",
          "osVersion": "11.2.1 (20D74)",
          "ownerType": "company",
          "partnerReportedThreatState": "unknown",
          "phoneNumber": "",
          "physicalMemoryInBytes": 0,
          "preferMdmOverGroupPolicyAppliedDateTime": "0001-01-01T00:00:00Z",
          "processorArchitecture": "unknown",
          "remoteAssistanceSessionErrorDetails": null,
          "remoteAssistanceSessionUrl": null,
          "requireUserEnrollmentApproval": true,
          "retireAfterDateTime": "0001-01-01T00:00:00Z",
          "roleScopeTagIds": Array [],
          "serialNumber": "[REDACTED]",
          "skuFamily": "",
          "skuNumber": 0,
          "specificationVersion": null,
          "subscriberCarrier": "",
          "totalStorageSpaceInBytes": 999653638144,
          "udid": null,
          "userDisplayName": "",
          "userId": "",
          "userPrincipalName": "",
          "usersLoggedOn": Array [],
          "wiFiMacAddress": "[REDACTED]",
          "windowsActiveMalwareCount": 0,
          "windowsRemediatedMalwareCount": 0,
        },
      },
    ],
    "_type": "intune_managed_device",
    "assetTag": "00000000-0000-0000-0000-000000000000",
    "category": "endpoint",
    "complianceState": "unknown",
    "createdOn": undefined,
    "deviceName": "[REDACTED]",
    "displayName": "[REDACTED]",
    "easDeviceId": false,
    "encrypted": true,
    "hardwareManufacturer": "Apple",
    "hardwareModel": "MacBook Pro",
    "hardwareSerial": "C02DG307MD6Q",
    "hardwareVendor": "Apple",
    "hardwareVersion": "MacBook Pro",
    "iccid": null,
    "id": "1f75329e-51fc-451c-bd40-c457337641a4",
    "imei": "",
    "jailBroken": true,
    "make": "Apple",
    "managed": true,
    "managementAgent": "mdm",
    "meid": "",
    "model": "MacBook Pro",
    "name": "[REDACTED]",
    "osDetails": "macOS version 11.2.1 (20D74)",
    "osName": "macOS",
    "osVersion": "11.2.1 (20D74)",
    "ownerType": "company",
    "phoneNumber": "",
    "physical": true,
    "registrationState": "registered",
    "serial": "[REDACTED]",
    "serialNumber": "[REDACTED]",
    "state": "retireIssued",
    "supervised": true,
    "udid": null,
    "userDisplayName": "",
    "userEmails": undefined,
    "userId": "",
    "username": "",
    "version": "MacBook Pro",
    "wifiMacAddress": "[REDACTED]",
  },
]
`;

exports[`fetchDevices Without Active Directory: managedDeviceRelationshipsNoAD 1`] = `
Array [
  Object {
    "_class": "HAS",
    "_key": "683dbff1-c8ff-4996-91ae-85484de46cfc|has|001d35fc-dd2b-4d17-9ffc-3bafa265e7d9",
    "_mapping": Object {
      "relationshipDirection": "REVERSE",
      "sourceEntityKey": "683dbff1-c8ff-4996-91ae-85484de46cfc",
      "targetEntity": Object {
        "_class": "User",
        "_key": "001d35fc-dd2b-4d17-9ffc-3bafa265e7d9",
        "_type": "azure_user",
        "email": "[REDACTED]",
        "id": "001d35fc-dd2b-4d17-9ffc-3bafa265e7d9",
      },
      "targetFilterKeys": Array [
        Array [
          "_class",
          "id",
        ],
        Array [
          "_class",
          "email",
        ],
      ],
    },
    "_type": "azure_user_has_intune_managed_device",
    "displayName": "HAS",
  },
]
`;

exports[`fetchDevices Without Active Directory: managedDevices 1`] = `
Array [
  Object {
    "BYOD": true,
    "_class": Array [
      "Device",
    ],
    "_key": "683dbff1-c8ff-4996-91ae-85484de46cfc",
    "_rawData": Array [
      Object {
        "name": "default",
        "rawData": Object {
          "aadRegistered": null,
          "activationLockBypassCode": null,
          "androidSecurityPatchLevel": "",
          "autopilotEnrolled": false,
          "azureADDeviceId": "683dbff1-c8ff-4996-91ae-85484de46cfc",
          "azureADRegistered": null,
          "azureActiveDirectoryDeviceId": "683dbff1-c8ff-4996-91ae-85484de46cfc",
          "chassisType": "unknown",
          "complianceGracePeriodExpirationDateTime": "2021-02-17T19:13:41Z",
          "complianceState": "noncompliant",
          "configurationManagerClientEnabledFeatures": null,
          "configurationManagerClientHealthState": null,
          "configurationManagerClientInformation": null,
          "deviceActionResults": Array [],
          "deviceCategoryDisplayName": "Unknown",
          "deviceEnrollmentType": "deviceEnrollmentManager",
          "deviceHealthAttestationState": null,
          "deviceName": "[REDACTED]",
          "deviceRegistrationState": "registered",
          "deviceType": "iPhone",
          "easActivated": false,
          "easActivationDateTime": "0001-01-01T00:00:00Z",
          "easDeviceId": "7CDT0MKDOH5FB0CKI5IVON0GJ8",
          "emailAddress": "[REDACTED]",
          "enrolledDateTime": "2021-02-17T16:24:45Z",
          "ethernetMacAddress": null,
          "exchangeAccessState": "none",
          "exchangeAccessStateReason": "none",
          "exchangeLastSuccessfulSyncDateTime": "0001-01-01T00:00:00Z",
          "freeStorageSpaceInBytes": 14481883136,
          "hardwareInformation": Object {
            "batteryChargeCycles": 0,
            "batteryHealthPercentage": 0,
            "batterySerialNumber": null,
            "cellularTechnology": null,
            "deviceFullQualifiedDomainName": null,
            "deviceGuardLocalSystemAuthorityCredentialGuardState": "running",
            "deviceGuardVirtualizationBasedSecurityHardwareRequirementState": "meetHardwareRequirements",
            "deviceGuardVirtualizationBasedSecurityState": "running",
            "freeStorageSpace": 0,
            "imei": "353249104017369",
            "ipAddressV4": null,
            "isEncrypted": false,
            "isSharedDevice": false,
            "isSupervised": false,
            "manufacturer": null,
            "meid": null,
            "model": null,
            "operatingSystemEdition": null,
            "operatingSystemLanguage": null,
            "operatingSystemProductType": 0,
            "osBuildNumber": null,
            "phoneNumber": null,
            "serialNumber": "C39ZQB9SN6XQ",
            "sharedDeviceCachedUsers": Array [],
            "subnetAddress": null,
            "subscriberCarrier": null,
            "totalStorageSpace": 0,
            "tpmSpecificationVersion": null,
            "wifiMac": null,
          },
          "iccid": null,
          "id": "683dbff1-c8ff-4996-91ae-85484de46cfc",
          "imei": "[REDACTED]",
          "isEncrypted": true,
          "isSupervised": false,
          "jailBroken": "False",
          "joinType": "unknown",
          "lastSyncDateTime": "2021-03-01T16:23:39Z",
          "lostModeState": "disabled",
          "managedDeviceName": "[REDACTED]",
          "managedDeviceOwnerType": "personal",
          "managementAgent": "mdm",
          "managementCertificateExpirationDate": "2022-02-15T19:56:28Z",
          "managementFeatures": "none",
          "managementState": "managed",
          "manufacturer": "Apple",
          "meid": "[REDACTED]",
          "model": "iPhone 11 Pro",
          "notes": null,
          "operatingSystem": "iOS",
          "osVersion": "14.4",
          "ownerType": "personal",
          "partnerReportedThreatState": "unknown",
          "phoneNumber": "[REDACTED]",
          "physicalMemoryInBytes": 0,
          "preferMdmOverGroupPolicyAppliedDateTime": "0001-01-01T00:00:00Z",
          "processorArchitecture": "unknown",
          "remoteAssistanceSessionErrorDetails": null,
          "remoteAssistanceSessionUrl": null,
          "requireUserEnrollmentApproval": null,
          "retireAfterDateTime": "0001-01-01T00:00:00Z",
          "roleScopeTagIds": Array [],
          "serialNumber": "[REDACTED]",
          "skuFamily": "",
          "skuNumber": 0,
          "specificationVersion": null,
          "subscriberCarrier": "Sprint",
          "totalStorageSpaceInBytes": 68719476736,
          "udid": null,
          "userDisplayName": "Michael Knoedel",
          "userId": "001d35fc-dd2b-4d17-9ffc-3bafa265e7d9",
          "userPrincipalName": "[REDACTED]",
          "usersLoggedOn": Array [
            Object {
              "lastLogOnDateTime": "2021-03-01T16:23:39.182Z",
              "userId": "001d35fc-dd2b-4d17-9ffc-3bafa265e7d9",
            },
          ],
          "wiFiMacAddress": "[REDACTED]",
          "windowsActiveMalwareCount": 0,
          "windowsRemediatedMalwareCount": 0,
        },
      },
    ],
    "_type": "intune_managed_device",
    "assetTag": "683dbff1-c8ff-4996-91ae-85484de46cfc",
    "category": "endpoint",
    "complianceState": "noncompliant",
    "createdOn": undefined,
    "deviceName": "[REDACTED]",
    "displayName": "[REDACTED]",
    "easDeviceId": false,
    "encrypted": true,
    "hardwareManufacturer": "Apple",
    "hardwareModel": "iPhone 11 Pro",
    "hardwareSerial": "C39ZQB9SN6XQ",
    "hardwareVendor": "Apple",
    "hardwareVersion": "iPhone 11 Pro",
    "iccid": null,
    "id": "683dbff1-c8ff-4996-91ae-85484de46cfc",
    "imei": "[REDACTED]",
    "jailBroken": false,
    "make": "Apple",
    "managed": true,
    "managementAgent": "mdm",
    "meid": "[REDACTED]",
    "model": "iPhone 11 Pro",
    "name": "[REDACTED]",
    "osDetails": "iOS version 14.4",
    "osName": "iOS",
    "osVersion": "14.4",
    "ownerType": "personal",
    "phoneNumber": "[REDACTED]",
    "physical": true,
    "registrationState": "registered",
    "serial": "[REDACTED]",
    "serialNumber": "[REDACTED]",
    "state": "managed",
    "supervised": false,
    "udid": null,
    "userDisplayName": "Michael Knoedel",
    "userEmails": Array [
      "[REDACTED]",
    ],
    "userId": "001d35fc-dd2b-4d17-9ffc-3bafa265e7d9",
    "username": "[REDACTED]",
    "version": "iPhone 11 Pro",
    "wifiMacAddress": "[REDACTED]",
  },
  Object {
    "BYOD": false,
    "_class": Array [
      "Device",
    ],
    "_key": "1f75329e-51fc-451c-bd40-c457337641a4",
    "_rawData": Array [
      Object {
        "name": "default",
        "rawData": Object {
          "aadRegistered": false,
          "activationLockBypassCode": null,
          "androidSecurityPatchLevel": "",
          "autopilotEnrolled": false,
          "azureADDeviceId": "00000000-0000-0000-0000-000000000000",
          "azureADRegistered": false,
          "azureActiveDirectoryDeviceId": "00000000-0000-0000-0000-000000000000",
          "chassisType": "unknown",
          "complianceGracePeriodExpirationDateTime": "0001-01-01T00:00:00Z",
          "complianceState": "unknown",
          "configurationManagerClientEnabledFeatures": null,
          "configurationManagerClientHealthState": null,
          "configurationManagerClientInformation": null,
          "deviceActionResults": Array [],
          "deviceCategoryDisplayName": "Unknown",
          "deviceEnrollmentType": "appleBulkWithoutUser",
          "deviceHealthAttestationState": null,
          "deviceName": "[REDACTED]",
          "deviceRegistrationState": "registered",
          "deviceType": "macMDM",
          "easActivated": false,
          "easActivationDateTime": "0001-01-01T00:00:00Z",
          "easDeviceId": "ApplC02DG307MD6Q",
          "emailAddress": "",
          "enrolledDateTime": "2021-02-17T16:38:03Z",
          "ethernetMacAddress": null,
          "exchangeAccessState": "none",
          "exchangeAccessStateReason": "none",
          "exchangeLastSuccessfulSyncDateTime": "0001-01-01T00:00:00Z",
          "freeStorageSpaceInBytes": 818191269888,
          "hardwareInformation": Object {
            "batteryChargeCycles": 0,
            "batteryHealthPercentage": 0,
            "batterySerialNumber": null,
            "cellularTechnology": null,
            "deviceFullQualifiedDomainName": null,
            "deviceGuardLocalSystemAuthorityCredentialGuardState": "running",
            "deviceGuardVirtualizationBasedSecurityHardwareRequirementState": "meetHardwareRequirements",
            "deviceGuardVirtualizationBasedSecurityState": "running",
            "freeStorageSpace": 0,
            "imei": "",
            "ipAddressV4": null,
            "isEncrypted": false,
            "isSharedDevice": false,
            "isSupervised": false,
            "manufacturer": null,
            "meid": null,
            "model": null,
            "operatingSystemEdition": null,
            "operatingSystemLanguage": null,
            "operatingSystemProductType": 0,
            "osBuildNumber": null,
            "phoneNumber": null,
            "serialNumber": "C02DG307MD6Q",
            "sharedDeviceCachedUsers": Array [],
            "subnetAddress": null,
            "subscriberCarrier": null,
            "totalStorageSpace": 0,
            "tpmSpecificationVersion": null,
            "wifiMac": null,
          },
          "iccid": null,
          "id": "1f75329e-51fc-451c-bd40-c457337641a4",
          "imei": "",
          "isEncrypted": true,
          "isSupervised": true,
          "jailBroken": "Unknown",
          "joinType": "unknown",
          "lastSyncDateTime": "2021-02-18T16:26:58Z",
          "lostModeState": "disabled",
          "managedDeviceName": "[REDACTED]",
          "managedDeviceOwnerType": "company",
          "managementAgent": "mdm",
          "managementCertificateExpirationDate": "2022-02-16T22:35:43Z",
          "managementFeatures": "none",
          "managementState": "retireIssued",
          "manufacturer": "Apple",
          "meid": "",
          "model": "MacBook Pro",
          "notes": null,
          "operatingSystem": "macOS",
          "osVersion": "11.2.1 (20D74)",
          "ownerType": "company",
          "partnerReportedThreatState": "unknown",
          "phoneNumber": "",
          "physicalMemoryInBytes": 0,
          "preferMdmOverGroupPolicyAppliedDateTime": "0001-01-01T00:00:00Z",
          "processorArchitecture": "unknown",
          "remoteAssistanceSessionErrorDetails": null,
          "remoteAssistanceSessionUrl": null,
          "requireUserEnrollmentApproval": true,
          "retireAfterDateTime": "0001-01-01T00:00:00Z",
          "roleScopeTagIds": Array [],
          "serialNumber": "[REDACTED]",
          "skuFamily": "",
          "skuNumber": 0,
          "specificationVersion": null,
          "subscriberCarrier": "",
          "totalStorageSpaceInBytes": 999653638144,
          "udid": null,
          "userDisplayName": "",
          "userId": "",
          "userPrincipalName": "",
          "usersLoggedOn": Array [],
          "wiFiMacAddress": "[REDACTED]",
          "windowsActiveMalwareCount": 0,
          "windowsRemediatedMalwareCount": 0,
        },
      },
    ],
    "_type": "intune_managed_device",
    "assetTag": "00000000-0000-0000-0000-000000000000",
    "category": "endpoint",
    "complianceState": "unknown",
    "createdOn": undefined,
    "deviceName": "[REDACTED]",
    "displayName": "[REDACTED]",
    "easDeviceId": false,
    "encrypted": true,
    "hardwareManufacturer": "Apple",
    "hardwareModel": "MacBook Pro",
    "hardwareSerial": "C02DG307MD6Q",
    "hardwareVendor": "Apple",
    "hardwareVersion": "MacBook Pro",
    "iccid": null,
    "id": "1f75329e-51fc-451c-bd40-c457337641a4",
    "imei": "",
    "jailBroken": true,
    "make": "Apple",
    "managed": true,
    "managementAgent": "mdm",
    "meid": "",
    "model": "MacBook Pro",
    "name": "[REDACTED]",
    "osDetails": "macOS version 11.2.1 (20D74)",
    "osName": "macOS",
    "osVersion": "11.2.1 (20D74)",
    "ownerType": "company",
    "phoneNumber": "",
    "physical": true,
    "registrationState": "registered",
    "serial": "[REDACTED]",
    "serialNumber": "[REDACTED]",
    "state": "retireIssued",
    "supervised": true,
    "udid": null,
    "userDisplayName": "",
    "userEmails": undefined,
    "userId": "",
    "username": "",
    "version": "MacBook Pro",
    "wifiMacAddress": "[REDACTED]",
  },
]
`;
